### ISSUE TITLE
Issue with multiple virtio-fs devices

### ISSUE THREAD

**Issue by rbradford** [cloud-hypervisor/cloud-hypervisor#3869]
When using two virtiofsd(C) instances and two --fs tags everything appears to be functioning nominally until:

login as user @ /home/user
firefox -headless

```
cloud-hypervisor: 13.514547634s: <_fs2> ERROR:virtio-devices/src/vhost_user/fs.rs:223 -- Failed to find RAM region associated with guest physical address 0xf2ede0100: InvalidGuestAddress(GuestAddress(65210810624))
cloud-hypervisor: 13.514629505s: <_fs2> ERROR:virtio-devices/src/vhost_user/mod.rs:272 -- Failed to handle request from vhost-user backend: ReqHandlerError(Os { code: 14, kind: Uncategorized, message: "Bad address" })
```

first fs tag is mounted as rootfs on /
second fs tag is mounted on /home/user

**The above error does not occur when running firefox -headless as root @ /root  which is the first fs tag**

Appears to be something wrong with how cloud hypervisor deals with multiple fs's.

_Originally posted by @jackloomen in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/3843#issuecomment-1072053610_

**Comment by jackloomen** [comment]
```
virtiofsd --socket-path=/path/rootfs.s -o source=/path/rootfs -o cache=auto
virtiofsd --socket-path=/path/home.s -o source=/path/home -o cache=auto
```


```
--cpus boot=4,max=4
--memory size=4G,shared=on
--fs tag=first,socket=/path/rootfs.s,dax=on,cache_size=1G tag=second,socket=/path/virtiofs/user.s,dax=on,cache_size=1G
cmdline += 'rootfstype=virtiofs root=first rootflags=dax'
```



guest init script:

`mount -t virtiofs -o dax second /home/user`

**Comment by sboeuf** [comment]
@jackloomen does it happen if you increase the cache_size to 8GiB for instance?

**Comment by jackloomen** [comment]
> @jackloomen does it happen if you increase the cache_size to 8GiB for instance?

with both tags having cache_size=8G the same issue occurs.

```
cloud-hypervisor: 6.334430947s: <_fs2> ERROR:virtio-devices/src/vhost_user/fs.rs:223 -- Failed to find RAM region associated with guest physical address 0xd731e0100: InvalidGuestAddress(GuestAddress(57765921024))
cloud-hypervisor: 6.33447997s: <_fs2> ERROR:virtio-devices/src/vhost_user/mod.rs:272 -- Failed to handle request from vhost-user backend: ReqHandlerError(Os { code: 14, kind: Uncategorized, message: "Bad address" })
```

**Comment by sboeuf** [comment]
Well on my side I'm having a very hard time getting `virtiofsd` (C) to work. It's always crashing, even when running everything as root (just to make sure I don't have permissions issues).
I'd actually rather try to build the latest version of `virtiofsd` (C) and make it work with CH before I go any further. My point is, some things might have been fixed since the version `qemu5.0-virtiofs-dax`.

**Comment by sboeuf** [comment]
Interestingly, the latest version of virtiofsd (C) (QEMU 7.0.0-rc0) works for me where the branch qemu5.0-virtiofs-dax wasn't working.

**Comment by sboeuf** [comment]
And for what it's worth, I can run Cloud Hypervisor exactly as you described (2 virtio-fs backends), the first one being the actual VM rootfs, and the second one being what can be found in /home/user, and nothing's failing for me, but I don't have firefox installed or any other program that might exercise the second virtio-fs backend.
Anyway I'll let you have a try with the latest QEMU version of virtiofsd (just in case use `sudo` to run both of your backends). And if you can provide me with the rootfs that you're using that'd be very helpful!

**Comment by jackloomen** [comment]
> Interestingly, the latest version of virtiofsd (C) (QEMU 7.0.0-rc0) works for me where the branch qemu5.0-virtiofs-dax wasn't working. I'm using the `main` branch from Cloud Hypervisor.

Where did you get QEMU 7.0.0-rc0? Is it virtio-fs-dev (https://gitlab.com/virtio-fs/qemu/-/tree/virtio-fs-dev)?
I doubt the rootfs matters, it was just alpine with firefox installed.

**Comment by jackloomen** [comment]
This was an extension of the previous virtiofs dax issue so it was not mentioned, but **this entire issue is with dax on and rootflag=dax mount -o dax.**

**Comment by sboeuf** [comment]
I was talking about the virtiofsd binary built from github.com/qemu/qemu.git

**Comment by jackloomen** [comment]
> I was talking about the virtiofsd binary built from github.com/qemu/qemu.git

I just tried it. Same results as virtio-fs-dev.

without rootflags=dax it boots and seems to be working but I'm not sure. Whenever I try to cat a file I get: cat: file: Function not implemented. Seems broken.

And with rootflags=dax it does not boot at all.

**Comment by sboeuf** [comment]
I'll do some more testing next week.

**Comment by sboeuf** [comment]
I had a try with the `virtio-fs-dev` branch from https://gitlab.com/virtio-fs/qemu.git and when using DAX (that is using `rootflags=dax`) I get the same error as you:
```
cloud-hypervisor: 307.616463ms: <_fs1> ERROR:virtio-devices/src/vhost_user/mod.rs:272 -- Failed to handle request from vhost-user backend: InvalidMessage
```
This is explained by the changes in the vhost-user protocol regarding the definition of `VhostUserFSSlaveMsg` structure. I have some patches which are fixing this issue locally. The main part of the change has to happen in `rust-vmm/vhost` as it must be updated according the latest set of patches for supporting DAX (https://gitlab.com/virtio-fs/qemu/-/commits/virtio-fs-dev).

That being said, the DAX support is still very experimental in the sense it hasn't been merged and is still very actively being updated. Therefore, I would suggest that you don't use it.

About the Rust version, can you summarize what kind of issues you were seeing?

**Comment by sboeuf** [comment]
Ok and if you don't use DAX, things are working?

**Comment by jackloomen** [comment]
Without DAX I observed no issues, other than the guest's uncontrollable usage of page cache - translating to elevated VMM memory consumption. But that's a kernel problem.

**Comment by sboeuf** [comment]
> Without DAX I observed no issues, other than the guest's uncontrollable usage of page cache - translating to elevated VMM memory consumption. But that's a kernel problem.
> 
> DAX off and cache=none solves that problem but breaks all applications using mmap on virtiofs mounts.

Yes I understand the blocker here. Problem is, DAX isn't stable so you can't really assume it's okay to use it.

I can continue the debugging with the Rust version if you can send me the rootfs you're using. I'm really trying to use the same conditions to reproduce the issue.

**Comment by jackloomen** [comment]
> > Without DAX I observed no issues, other than the guest's uncontrollable usage of page cache - translating to elevated VMM memory consumption. But that's a kernel problem.
> > DAX off and cache=none solves that problem but breaks all applications using mmap on virtiofs mounts.
> 
> Yes I understand the blocker here. Problem is, DAX isn't stable so you can't really assume it's okay to use it.
> 
> I can continue the debugging with the Rust version if you can send me the rootfs you're using. I'm really trying to use the same conditions to reproduce the issue.

Alpine rootfs.
apk add firefox
boot rootflags=dax
env -i -- firefox -headless

If you don't get virtiofsd(rust) to panic then I would very much want to know what configuration you are using. I tried 3 different rootfs' (glibc and musl) and all had the same result.

**Comment by jackloomen** [comment]
There's also:
https://gitlab.com/virtio-fs/virtiofsd/-/issues/39

Problem wasn't verified on qemu.

**Comment by sboeuf** [comment]
I've been able to reproduce the issue consistently with an alpine rootfs to which I added firefox. I get the exact same error.
I identified the problem, it's related to the address found in the VIRTIO descriptor. Whenever this address isn't referring to the guest memory, it should be handled with a special command `VHOST_USER_SLAVE_FS_IO` so that we ask the VMM to perform the memory access on behalf of the `virtiofsd` backend.
This functionality is implemented in the latest C version (which I have working locally with the vhost-user protocol patches applied), but it doesn't in the Rust version since this is not implemented.
The Rust version of `virtiofsd` should be patched, otherwise we can't consider it as supporting DAX at all.

**Comment by sboeuf** [comment]
Let's follow up in there: https://gitlab.com/virtio-fs/virtiofsd/-/issues/41
